p2p/proto/snapshot.proto

```diff
@@ -0,0 +1,105 @@
+syntax = "proto3";
+
+import "p2p/proto/common.proto";
+import "p2p/proto/state.proto";
+
+message PatriciaNode {
+    message Edge {
+        uint32  length = 1;
+        Felt252 path   = 2;  // as bits of left/right
+        Felt252 value  = 3;
+    }
+    message Binary {
+        Felt252 left  = 1;
+        Felt252 right = 2;
+    }
+
+    oneof node {
+        Edge   edge   = 1;
+        Binary binary = 2;
+    }
+}
+
+// non leaf nodes required to build the trie given the range (leaves)
+message PatriciaRangeProof {
+    repeated PatriciaNode nodes = 1;
+}
+
+// leafs of the contract state tree
+message ContractState {
+    Hash   address = 1;  // the key
+    Hash   class   = 2;
+    Hash   storage = 3;  // patricia
+    uint64 nonce   = 4;
+}
+
+// request a range from the contract state tree that matches the given root (block)
+// starts at 'start' and ends no more than 'end'.
+// the result is  (ContractRange+, PatriciaRangeProof)*
+message ContractRangeRequest {
+    uint32 domain           = 1;  // volition
+    Hash   state_root       = 2;
```

Comment:
Don't forget, you'll need to have a way to get the root of the contract tree and the class tree. I see only one state root hash on `BlockHeader` so you'll need to add another endpoint to get the two root.

Reply:
There is ClassRangeRequest

Reply:
I mean, when using the proof to verify the integrity of the message, its going to need to compare the calculated tree root with a known root. Since the state hash in the header is the hash of the [two tree root](https://docs.starknet.io/documentation/architecture_and_concepts/Network_Architecture/starknet-state/#state_commitment) instead of the root itself, the client don't know what to compare the generated tree root with.

Reply:
Correct.

Reply:
I will add the two hashes to ContractRangeResponse and ClassRangeResponse. It's 64 bytes addition to a 1M message.

Reply:
Yea, I recommend another endpoint for that. See [this](https://github.com/NethermindEth/juno/pull/1030/files#diff-a04fd2fd0785454e284c95dea53642c1d266c8cbbf66a8862b330362cf710e87R45). The two root itself need to be verified by recalculating them to get back the block state root. Since it need both to calculate back the hash in header, its easier to have it in one response.

Reply:
Instead of another endpoint, I added the two hashes to each response. This is just to save on another round trip.

Reply:
That'll work too.